File: JPA/sample-projects/shop-api/src/main/java/me/kzv/shopapi/domain/OrderItem.java
package me.kzv.shopapi.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import me.kzv.shopapi.domain.item.Item;

import javax.persistence.*;

@Entity
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED) // protected 로 생성자를 만들어 다른 클래스에서 바로 생성하여 사용할 수 없게 만든다.
public class OrderItem {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "order_item_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    private Item item;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    private Order order;

    private int orderPrice; // 주문 가격
    private int count; // 주문 수량

    //==생성 메서드==//
    public static OrderItem createOrderItem(Item item, int orderPrice, int count) {
        OrderItem orderItem = new OrderItem();
        orderItem.setItem(item);
        orderItem.setOrderPrice(orderPrice);
        orderItem.setCount(count);

        item.removeStock(count); // 재고 감소
        return orderItem;
    }

    //==비즈니스 로직==//
    /**
     * 주문 취소 - 재고 수량 원복
     */
    public void cancel(){
        getItem().addStock(count);
    }

    /**
     * 주문상품 전체 가격 조회
     */
    public int getTotalPrice(){
        return getOrderPrice() * getCount();

    }
}

File: JPA/kotlin-jpa-querydsl/src/main/kotlin/me/kzv/kotlinjpaquerydsl/entity/OrderItem.kt
package me.kzv.kotlinjpaquerydsl.entity

import jakarta.persistence.*
import me.kzv.kotlinjpaquerydsl.entity.items.Item

@Entity
class OrderItem(
    @Id
    @GeneratedValue
    val id: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    var item: Item,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    var order: Order? = null, // orderItem 이 먼저 생성된 후 order 가 생성되므로 일단은 null

    val orderPrice: Int,
    val count: Int,
) {
    //==생성 메서드==//
    companion object {
        fun createOrderItem(item: Item, orderPrice: Int, count: Int): OrderItem {
            val orderItem = OrderItem(item = item, orderPrice = orderPrice, count = count)
            item.removeStock(count)
            return orderItem
        }
    }

    //==비즈니스 로직==//
    /**
     * 주문 취소 - 재고 수량 원복
     */
    fun cancel() {
        item.addStock(count)
    }

    /**
     * 주문상품 전체 가격 조회
     */
    fun getTotalPrice(): Int = orderPrice * count

}
